### DISCUSSION TITLE
How do cmdlets return multiple objects as an array?

### DISCUSSION THREAD

**Discussion by mcdonamw** [PowerShell/PowerShell#18601] [Q&A]
**###EDIT###  Rewriting this post to better clarify my ask.**

I had an issue in the past where I was working on some code where I was creating a large dynamic collection (1million+) and was doing so with an array and use of the += method to add entries to it (because I did not know any better!).  When I had 1000 or so it was relatively quick but when I released it to the full dataset, it then took over 6 hours!!!

Through research, I found that using Arrays and the += method is a nightmare, especially as it gets larger because it literally recreates the array every time a new addition occurs.

Ultimately I fixed my issue by using an ArrayList instead, which then made my code run in mere seconds.  That huge discrepancy in time shocked me.  But I digress.

What really stumped me was that I know many built in cmdlets in Powershell returns collections as arrays, and does so very quickly.  For example:

```
PS C:\> $p = Get-Process
PS C:\> $p.GetType()

IsPublic IsSerial Name                                     BaseType
-------- -------- ----                                     --------
True     True     Object[]                                 System.Array
```

Now, obviously this cmdlet would never take a long time because there will never really be that many processes.  I'm just using this as an example to show the type of the collection it spits out.  

**With that said,** I'm betting that if I tried to reproduce these results by building my own function to simply read these processes from a file, and build a collection of type Array, it would take much much longer than what it does natively.  _The file read is negligible here as it would be instant._  

So that leads me to ask myself:  What in the world is Powershell doing in the backend to create large indeterminate collections that are ultimately spit out as arrays?  It's clearly not populating an empty array from the start, though it ultimately outputs one.  

With all that said, as I solved my original issue, I did find that an ArrayList has a method called ToArray() that will convert said arraylist to an array, and it's near instant, even with my million+ item ArrayList.   So I'm wondering if Powershell does something similar.  Like, is it building an arraylist in the background and then simply running ToArray() after?  

I don't imagine any of this is documented.  I'm just wondering if anyone by chance knows.

**Comment by ThomasNieto** [comment]
PowerShell cmdlets send objects one at a time down the pipeline. This allows for pseudo multi-threading as while the first cmdlet is still outputting objects to the pipeline the second command can start working on the results. Objects that reach the end of the pipeline are returned as a single array. For more information see: [about_pipelines](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_pipelines?view=powershell-7.3#one-at-a-time-processing)

As you found out trying to use `+=` on an array while cause significant performance issues the larger the array gets because the array is recreated each time. If you're writing a cmdlet/function that needs to output multiple items use the pipeline, because then you don't need to manage the collection of objects. Here is some more information about the performance implications of the various collection types: https://powershell.one/tricks/performance/arrays

```powershell
Get-Process | Stop-Process -WhatIf
```

To see the the objects being sent one at a time we put a delay in the pipeline.

```powershell
Get-Process | ForEach-Object { $_; Start-Sleep 2 } | Stop-Process -WhatIf
```

**Reply by mcdonamw** [reply]
I apologize for the delay, but I lost track of this post entirely.  

I fully understand how Powershell works on the pipeline.  That's not my question.  I have reworked my original post to better clarify what I'm asking.

**Comment by Northman-de** [comment]
Using the pipeline is not always possible; e.g. creating multiple arrays during a loop.
This script demonstrates the negative effect on the runtime. It adds elements to an array using += and outputs the milliseconds needed to add the last 1000 items:
```powershell
$array = @()
$datetime = Get-Date
for ($i = 0 ; $i -lt 100000 ; $i++) {
	$array += $i % 100
	if ($i % 1000 -eq 0) {
		"{0,6:0} - {1,10:0}" -f $i,((Get-Date) - $datetime ).Totalmilliseconds
		$datetime = Get-date
	}
}
```
Output:
```
  1000 -         30
  2000 -         67
  3000 -        115
  4000 -        146
  5000 -        185
  6000 -        236
  7000 -        273
  8000 -        313
  9000 -        525
 10000 -        507
...
 95000 -       5698
 96000 -       5682
 97000 -       5697
 98000 -       5902
 99000 -       5876
```
Again: the second number is NOT the total runtime. It's the time needed for the last 1000 iterations. The script runs for 5 minutes on my computer.

Use arraylists instead. They are much faster:
```powershell
$array = New-Object -TypeName System.Collections.ArrayList
$datetime = Get-Date
for ($i = 0 ; $i -lt 100000 ; $i++) {
	[Void]$array.Add($i % 100)
	if ($i % 1000 -eq 0) {
		"{0,6:0} - {1,10:0}" -f $i,((Get-Date) - $datetime ).Totalmilliseconds
		$datetime = Get-date
	}
}
```
The complete loop needs only ~100 milliseconds. So its  5min * 60sec / 0,1s ~ 3,000 times faster.
A loop to 1,000,000 without the output needs 300ms.
The arraylists can be used like a normal array (foreach / export-csv).

**Reply by mcdonamw** [reply]
Apologies for the delay.  I lost track of this post.

But yes, you are correct.  I ended up solving my original issue by doing this very thing;  using an ArrayList.  That's not what I'm asking here.  Please revisit the post as I reworked it to better clarify what I'm asking.

**Comment by powercode** [comment]
see https://learn.microsoft.com/en-us/dotnet/api/system.management.automation.cmdlet.writeobject?view=powershellsdk-7.3.0

The WriteObject method has two overloads, one which sends arrays to the pipeline as an array, and another that has a boolean parameter that controls weather a collection should be enumerated or not.

| Method | Description |
|---|--|
| WriteObject(Object) | Writes the object to the output pipe. |
| WriteObject(Object, Boolean) | Writes one or more objects to the output pipe. If the object is a collection and the enumerateCollection flag is true, the objects in the collection will be written individually. |

**Reply by mcdonamw** [reply]
Thanks for your input, but this does not seem to address my question.  I have reworked the original post to better clarify.

**Comment by Buzman01** [comment]
A cmdlet which returns multiple objects, what collection type is it, if any?

An example of the Get-ADuser cmdlet:

$Users = Get-ADuser -Filter * 

It will in most cases return multiple ADuser objects, but what "collection" type is it? The documentation only says it will returns one or more user objects of Microsoft.ActiveDirectory.Management.ADUser.

Tried to use e.g. ($Users -is [System.Collections.ArrayList]) but I cannot nail the "collection" type?

arrays

powershell

collections

powershell-cmdlets

Cmdlets themselves typically use no collection type in their output.[1]:
They emit individual objects to the pipeline, which can situationally mean: zero, one, or multiple ones.

This is precisely what Get-ADUser does: the specific number of output objects depends on the arguments that were given; that is why the Get-AdUser help topic only mentions scalar type ADUser as the output type and states that it "returns one or more" of them.

Generally, the PowerShell pipeline is designed to be a conduit for a stream of objects, whose count need not be known in advance, with commands typically outputting objects one by one, as soon as they become available, and with receiving commands typically also processing them one by one, as soon as they're being received. (see about_Pipelines).

It is the PowerShell engine itself that automatically collects multiple outputs for you in an [object[]] array[2] if needed, notably if you capture output via a variable assignment or use a command call via (...), the grouping operator (or $(...), the subexpression operator[3], or @(...), the array subexpression operator, discussed in detail below), as an expression:

# Get-ChildItem C:\Windows has *multiple* outputs, so PowerShell # collects them in an [object[]] array. PS> $var = Get-ChildItem C:\Windows; $var.GetType().Name Object[] # Ditto with (...) (and also with $(...) and always with @(...)) PS> (Get-ChildItem C:\Windows).GetType().Name Object[] 

However, if a given command - possibly situationally - only outputs a single object, you'll then get just that object itself - it is not wrapped in an array (unless you use @(...) - see below):

# Get-Item C:\ (always) returns just 1 object. PS> $var = Get-Item C:\; $var.GetType().Name DirectoryInfo # *not* a single-element array, # just the System.IO.DirectoryInfo instance itself 

What can get tricky is that a given command can situationally produce either one or multiple outputs, depending on inputs and runtime conditions, so the engine may return either a single object or an array.

# !! What $var receives depends on the count of subdirs. in $HOME\Projects: PS> $var = Get-ChildItem -Directory $HOME\Documents; $var.GetType().Name ??? # If there are *2 or more* subdirs: an Object[] array of DirectoryInfos. # If there is only *one* subdir.: a DirectoryInfo instance itself. # (See below for the case when there is *no* output.) 

@(...), the array-subexpression operator, is designed to eliminate this ambiguity, if needed: By wrapping a command in @(...), PowerShell ensures that its output is always collected as [object[]] - even if the command happens to produce just one output object or even none:

PS> $var = @(Get-ChildItem -Directory $HOME\Projects); $var.GetType().Name Object[] # Thanks to @(), the output is now *always* an [object[]] array. 

With variable assignments, a potentially more efficient alternative is to use an [array] type constraint to ensure that the output becomes an array:

# Alternative to @(...) # Note: You may also create a strongly typed array, with on-demand type conversions: # [string[]] $var = ... PS> [array] $var = Get-ChildItem -Directory $HOME\Documents; $var.GetType().Name Object[] 

Note:

This is potentially more efficient in that if the RHS already happens to be an array, it is assigned as-is, whereas @(...) actually enumerates the output from ... and then reassembles the elements into a new ([object[]]) array.

[array] preserves the specific type of an input array by simply passing it through (e.g., [array] $var = [int[]] (1..3) stores the [int[]] array as-is in $var).

Placing the [array] "cast" to the left of $var = ... - which is what it makes it a type constraint on the variable - means that the type of the variable is locked in, and assigning different values to $var later will continue to convert the RHS value to [array] ([object[]]), if needed (unless you assign $null or "nothing" (see below)).

Taking a step back: Ensuring that collected output is an array is often not necessary, due to PowerShell's unified handling of scalars and collections in v3+:

PowerShell exposes intrinsic members even on scalar (non-collection) objects that allow you treat them like collections.

E.g. (42).Count is 1 ((42).Length works too); it is treated like an integer collection with 1 element, and this also applies to indexing: (42)[0] is 42, i.e. the virtual collection's first and only element.

Caveat: Type-native members take precedence, which is a pitfall with strings: 'foo'.Length is 3, the string's length - but 'foo'.Count works (is 1); however, 'foo'[0] is invariably 'f', because the [string] type's native indexer returns the individual characters in the array (workaround: @('foo')[0])

See this answer for details.

If a command produces no output, you'll get "nothing" (strictly speaking: the [System.Management.Automation.Internal.AutomationNull]::Value singleton), which in most cases behaves like $null[4]:

# Get-Item nomatchingfiles* produces *no* output. PS> $null -eq (Get-Item nomatchingfiles*) True # Conveniently, PowerShell lets you call .Count on this value, which the # behaves like an empty collection and indicates 0. PS> (Get-Item nomatchingfiles*).Count 0 

[1] It is possible to output entire collections as a whole to the pipeline (in PowerShell code with Write-Output -NoEnumerate $collection or, more succinctly, , $collection), but that is then just another object in the pipeline that happens to be a collection itself. Outputting collections as a whole is an anomaly, however, that changes how commands you pipe to see the output, which can be unexpected; a prominent example is ConvertFrom-Jsons unexpected behavior prior to v7.0.

[2] a System.Array instance whose elements are of type System.Object, allowing you to mix objects of different types in a single array.

[3] Use of (...) is usually sufficient; $(...) is only needed for string interpolation (expandable strings) and for embedding whole statements or multiple commands in a larger expression; note that $(...), unlike (...) by itself, unwraps single-element arrays; compare (, 1).GetType().Name to $(, 1).GetType().Name; see this answer.

[4] There are scenarios in which "nothing" behaves differently from $null, notably in the pipeline and in switch statements, as detailed in this comment on GitHub; the linked issue is a feature request to make "nothing" more easily distinguishable from $null, by supporting -is [AutomationNull] as a test.

**Reply by mcdonamw** [reply]
I appreciate your input here, but with respect, this is far too much information and I don't think it really touches on what I'm asking.

I'll rewrite my original post to clarify.

**Comment by powercode** [comment]
I think you will get an incorrect model if you think of it as "returning". 

Functions and cmdlets write objects to the pipeline, where they can be read by other cmdlets or captured in variables.
When you assign the result of a pipeline to a variable, it will become an `object[]` if there are multiple objects written to the pipeline and the type of the item if it is a single object written to the pipeline.

Consider the following:

```powershell
PS> $multi = 1..10
PS> Get-Member -InputObject $multi
```
```text
  TypeName: System.Object[]
  ...
```

```powershell
PS> $scalar = 1..10 | Select-Object -First 1
PS> Get-Member -InputObject $scalar
```
```text
  TypeName: System.Int32
```

To force the result of a pipeline to always be a collection, you can do
```powershell
PS> $multi = @(1..10 | Select-Object -First 1)
PS> Get-Member -InputObject $multi
```
```text
  TypeName: System.Object[]
```


As a side note: @Buzman01 I would strongly urge upon you to spend some time reading up on how to format with markdown. Otherwise, your text becomes hard to read and made me only read snippets of it as it felt like a wall of text.

**Reply by Buzman01** [reply]
Oh thanks so much for letting me know

**Reply by powercode** [reply]
The small M with a down-arrow in the lower right corner is a link to the syntax

**Reply by mcdonamw** [reply]
Appreciate your input, but my inquiry has nothing to do with pipelines.  I will rewrite my post to better clarify.

**Comment by powercode** [comment]
Yes. PowerShell has a `PSDataCollectionStream<T>` that wraps a `PSDataCollection<T>`.
See `src\System.Management.Automation\utils\ObjectStream.cs`.

Making a copy of a collection into an array is quite fast. The slowdown comes from the fact that the `+=` operator on an array creates a new array, with size+1, copying the source array and adding a last element.
This happens for every element you add, so if you add a million elements, you make a million array copies, of bigger and bigger arrays.
